APDS-61 - [BE] add API to list and delete permanent tokens

poxip opened the pull request:
This also closes APDS-60. 

In this PR I had to drop support for DRF 3.1 and 3.2, because their code is incompatible with Django >= 1.10 (see: https://github.com/encode/django-rest-framework/issues/3252 - it has been fixed in 3.3). I can't do anything about it to make it backwards compatible.

poxip committed:
APDS-61 - add API to list and delete permanent tokens

remik commented on rest_framework_jwt/views.py:
```diff
@@ -99,6 +101,23 @@ class RefreshJSONWebToken(JSONWebTokenAPIView):
+    def get_queryset(self):
+        return self.queryset.filter(user=self.request.user)
+
+    def get(self, *args, **kwargs):
```
Why do we need this and below if you have mixins?

poxip replied:
Mixins used here only add the "destroy" and "list" methods

remik replied:
We used mixins so router should discover it. Check: http://www.django-rest-framework.org/api-guide/viewsets/

remik commented on tests/test_views.py:
```diff
@@ -492,3 +495,38 @@ def test_refresh_jwt_after_refresh_expiration(self):
         api_settings.JWT_ALLOW_REFRESH = DEFAULTS['JWT_ALLOW_REFRESH']
+
+
+class DeviceViewTests(TokenTestCase):
```
Please add devices for not your user to make sure we are not showing it and can't delete.

poxip replied:
Added in the CR commit.

remik commented on tests/urls.py:
```diff
@@ -28,6 +28,8 @@ def post(self, request):
     url(r'^auth-token/$', views.obtain_jwt_token),
     url(r'^auth-token-refresh/$', views.refresh_jwt_token),
     url(r'^auth-token-verify/$', views.verify_jwt_token),
+    url(r'^devices/(?P<pk>[\w]+)$', views.device_delete_view),
```
Use router.

poxip committed:
CR

poxip committed:
Use router

poxip committed:
Revert tox.ini

poxip commented:
The tests are okay (see details), Travis does not provide Python3.3 somehow (it works in the original repo though). We do not have to drop support for DRF 3.0, 3.1, 3.2 because those versions are only supposed to work with Django 1.8 and 1.9.

poxip committed:
Revert Readme

poxip committed:
Use format='json' in requests

poxip committed:
Modify client.login() call

poxip committed:
Add auto_now to Device.last_request_datetime

poxip committed:
Revert .travis.yml

poxip committed:
Tox: remove py33 and add py36

poxip committed:
tox: revert and add Python 3.3

poxip committed:
Do not test under Python 3.3

jacoor commented on rest_framework_jwt/models.py:
```diff
@@ -1 +1,16 @@
+    user = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE)
+    name = models.CharField(_('Device name'), max_length=255)
+    details = models.CharField(_('Device details'), max_length=255, blank=True)
+    last_request_datetime = models.DateTimeField(auto_now=True)
```
Missing empty line at the end of file.
Pylint and pep8 will complain.

poxip replied:
GitHub does not show the last empty line as far as I remember. The flake8 task in tox did not complain, see: https://travis-ci.org/ArabellaTech/django-rest-framework-jwt/jobs/256993179

jacoor commented on rest_framework_jwt/serializers.py:
```diff
@@ -169,3 +170,9 @@ def validate(self, attrs):
+class DeviceSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Device
+        fields = ['permanent_token', 'jwt_secret', 'created', 'name', 'details', 'last_request_datetime']
```
WHat?! you return JWT secret to the world? 
http://getblimp.github.io/django-rest-framework-jwt/#jwt_secret_key Please check documentation what JWT secret is.

jacoor commented on tests/test_views.py:
```diff
@@ -492,3 +493,46 @@ def test_refresh_jwt_after_refresh_expiration(self):
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(len(response.data), 1)
+        self.assertEqual(set(response.data[0].keys()), {
+            'permanent_token', 'jwt_secret', 'created', 'name', 'details', 'last_request_datetime'
```
Please, do not make any field *_secret_* public. After all, there is a reason why they call it secret.
http://getblimp.github.io/django-rest-framework-jwt/#jwt_secret_key

poxip committed:
CR

Change jwt_secret to UUUIDField with auto value, add Device.id and remove vulnerable data from Device serializer

poxip committed:
APDS-65 - [BE] modify login API to generate permanent token (#2)

* APDS-65 - [BE] modify login API to generate permanent token

* Remove TODO

* Remove python select from travis config

* Use user agent

* Remove Python 3.3 from travis

* CR

* Correct

* APDS-64 - [BE] add API to logout device (#6)

* APDS-64 - [BE] remove Device on logout

* Corrects

* Replace doublequotes with singlequotes

* Remove duplicated lines

* Remove python setting from Travis config

* Use user agent in logout view

* Fix tests

* Remove Python 3.3 from travis

* CR

* CR

* APDS-63, APDS-62 - [BE] add API to refresh token using permanent_token (#4)

* APDS-63 - [BE] disallow passing permanent_token header in views other than device refresh

This requires APDS-62 to be finished

* APDS-62 - [BE] add API to refresh token using permanent_token

* Replace doublequotes with singlequotes

* Update with changes from APDS-62

* Remove python setting from .travis.yml

* Correct tests

* Auto logout when permanent token has expired

* Remove Python 3.3 from travis

* CR

* Remove addons from .travis.yml
